homeassistant/components/denonavr/media_player.py

```diff
@@ -524,3 +569,58 @@ async def async_set_dynamic_eq(self, dynamic_eq: bool):
 
         if self._update_audyssey:
             await self._receiver.async_update_audyssey()
+
+    @async_log_errors
+    async def async_send_telnet_commands(self, commands: str) -> list[dict[str, str]]:
+        """Send telnet commands to receiver and subscribe to its responses."""
+        command_list = commands.splitlines()
+        events = []
+        for command in command_list:
+            found = False
+            for event in TELNET_EVENTS_SORTED:
+                if command.startswith(event):
+                    events.append(event)
+                    found = True
+                    break
+            if not found:
+                raise HomeAssistantError(
+                    f"Invalid command {command}. Valid commands start with these prefixes: {TELNET_EVENTS_SORTED}"
+                )
+
+        result = []
+        times = {"last_message": 0.0}
+
+        async def async_telnet_callback(zone: str, event: str, parameter: str) -> None:
+            times["last_message"] = time.monotonic()
+            if zone in [MAIN_ZONE, self._receiver.zone]:
+                result.append({"event": event, "parameter": parameter})
+
+        for event in events:
+            self._receiver.register_callback(event, async_telnet_callback)
+
+        times["start"] = time.monotonic()
+        success = self._receiver.send_telnet_commands(*command_list)
+        if success:
+            while True:
+                if time.monotonic() - times["start"] > DEFAULT_TIMEOUT:
+                    _LOGGER.debug("Telnet command returns because timeout elapsed")
+                    break
+                if (
+                    times["last_message"]
+                    and time.monotonic() - times["last_message"] > 0.2
+                ):
+                    _LOGGER.debug(
+                        "Telnet command returns because last message timeout elapsed"
+                    )
+                    break
+                await asyncio.sleep(0.1)
```

Comment:
Maybe wrap this in a `try` and do the unregister in `finally` in case anything fails it doesn't end up leaking a registration